File: kirun-js/src/engine/dsl/parser/ast/ASTNode.ts
import { SourceLocation } from '../../lexer/DSLToken';

/**
 * Base class for all AST nodes
 */
export abstract class ASTNode {
    constructor(
        public readonly type: string,
        public readonly location: SourceLocation,
    ) {}

    /**
     * Convert AST node to JSON (for debugging/inspection)
     */
    abstract toJSON(): any;

    /**
     * Pretty-print the AST node
     */
    public toString(): string {
        return JSON.stringify(this.toJSON(), null, 2);
    }
}

File: kirun-py/src/kirun_py/dsl/parser/ast/ast_node.py
from __future__ import annotations

import json
from abc import ABC, abstractmethod
from typing import Any

from kirun_py.dsl.lexer.dsl_token import SourceLocation


class ASTNode(ABC):
    """Base class for all AST nodes."""

    def __init__(self, type: str, location: SourceLocation) -> None:
        self.type = type
        self.location = location

    @abstractmethod
    def to_json(self) -> Any:
        """Convert AST node to JSON (for debugging/inspection)."""
        ...

    def __str__(self) -> str:
        """Pretty-print the AST node."""
        return json.dumps(self.to_json(), indent=2)

    def __repr__(self) -> str:
        return self.__str__()
